### PR TITLE
feat: flash liquidation proto

### PR DESCRIPTION
**Pull request by toteki**
Proto (and scaffolding around generated framework) only in this PR.

### CODE REVIEW

File: x/leverage/types/tx.go [hunk 1]
```diff
@@ -255,6 +255,39 @@ func (msg *MsgLiquidate) GetSignBytes() []byte {
...
+func NewMsgFlashLiquidate(liquidator, borrower sdk.AccAddress, repayDenom, rewardDenom string) *MsgFlashLiquidate {
+	return &MsgFlashLiquidate{
+		Liquidator:  liquidator.String(),
+		Borrower:    borrower.String(),
+		RepayDenom:  repayDenom,
+		RewardDenom: rewardDenom,
+	}
+}
+
+func (msg MsgFlashLiquidate) Route() string { return sdk.MsgTypeURL(&msg) }
```
Comment (gsk967) [inline x/leverage/types/tx.go:267]:
```suggestion
func (msg MsgFlashLiquidate) Route() string { return "" }

```

File: proto/umee/leverage/v1/tx.proto [hunk 2]
```diff
@@ -140,6 +149,21 @@ message MsgLiquidate {
   string reward_denom = 4;
 }
 
+// MsgFlashLiquidate is the request structure for the FlashLiquidate RPC.
+message MsgFlashLiquidate {
+  // Liquidator is the account address performing a liquidation and the signer
+  // of the message.
+  string liquidator = 1;
```
Comment (gsk967) [inline proto/umee/leverage/v1/tx.proto:156]:
```suggestion
  string liquidator = 1 [(cosmos_proto.scalar) = "cosmos.AddressString"];
```

File: proto/umee/leverage/v1/tx.proto [hunk 3]
```diff
@@ -140,6 +149,21 @@ message MsgLiquidate {
...
+  // Borrower is the account whose borrow is being repaid, and collateral consumed,
+  // by the liquidation. It does not sign the message.
+  string borrower = 2;
```
Comment (gsk967) [inline proto/umee/leverage/v1/tx.proto:159]:
```suggestion
  string borrower = 2 [(cosmos_proto.scalar) = "cosmos.AddressString"];
```

**Review by gsk967 — APPROVED** [review]
LGTM
I left some comments, Please check those

File: proto/umee/leverage/v1/tx.proto [hunk 4]
```diff
@@ -140,6 +149,21 @@ message MsgLiquidate {
...
+  // RepayDenom is the base token that the liquidator will borrow and repay on behalf of
+  // the borrower.
+  string repay_denom = 3;
+  // RewardDenom is the uToken denom that the liquidator will receive as a liquidation reward.
+  string reward_denom = 4;
+}
```
[the remark this answers is not in the document]
Comment (toteki) [reply proto/umee/leverage/v1/tx.proto:165]: In the current approach, users don't calculate anything (input is denoms only)

But I can look into making the health chech at the end of the transaction more advanced - perhaps requiring <90% borrow limitto succeed

File: proto/umee/leverage/v1/tx.proto (hunk 3 above)
Comment (toteki) [reply proto/umee/leverage/v1/tx.proto]: added

File: proto/umee/leverage/v1/tx.proto [hunk 5]
```diff
@@ -48,6 +48,15 @@ service Msg {
...
   rpc Liquidate(MsgLiquidate) returns (MsgLiquidateResponse);
 
+  // FlashLiquidate allows a user to repay a different user's borrowed coins in exchange for some
+  // of the target's collateral. For flash liquidations, the tokens to repay are borrowed instead of
+  // being taken from the liquidator's wallet, and the reward is immediately collateralized. Borrow
+  // limit checks for the liquidator are deferred until after the reward is collateralized, allowing
+  // this initial borrow to exceed the liquidator's borrow limit as long as it is healthy by the end
+  // of the transaction. Repay amount is calculated automatically, so the liquidator only specifies
+  // repay and reward token denoms.
+  rpc FlashLiquidate(MsgFlashLiquidate) returns (MsgFlashLiquidateResponse);
```
[the remark this answers is not in the document]
Comment (toteki) [reply proto/umee/leverage/v1/tx.proto]: Renamed to FastLiquidate for now to avoid "flash"

**Pull request merged**